src/frontend/src/binder/expr/binary_op.rs

```diff
@@ -71,26 +71,22 @@ impl Binder {
 
     /// Bind `||`. Based on the types of the inputs, this can be string concat or array concat.
     fn bind_concat_op(&mut self, left: ExprImpl, right: ExprImpl) -> Result<ExprImpl> {
-        let types = [left.return_type(), right.return_type()];
-
-        let has_string = types.iter().any(|t| matches!(t, DataType::Varchar));
-        let has_array = types.iter().any(|t| matches!(t, DataType::List { .. }));
-
-        // StringConcat
-        if has_string && !has_array {
-            Ok(FunctionCall::new(ExprType::ConcatOp, vec![left, right])?.into())
-        }
-        // ArrayConcat
-        else if has_array {
-            Err(ErrorCode::NotImplemented("array concat operator".into(), None.into()).into())
-        }
-        // Invalid types
-        else {
-            Err(ErrorCode::BindError(format!(
-                "operator does not exist: {:?} || {:?}",
-                &types[0], &types[1]
-            ))
-            .into())
-        }
+        let func_type = match (left.return_type(), right.return_type()) {
+            // string concatenation
+            (DataType::Varchar, _) | (_, DataType::Varchar) => ExprType::ConcatOp,
```

Comment:
As we use separate expression types for arrays, I think it's okay to name this more specific like `StringConcatOp`. 😄